src/twisted/newsfragments/11883.feature

```diff
@@ -0,0 +1 @@
+Twisted currently doesn't work at all when openssl fips is enabled in the host due to md5 algorithm usage in various places. This patch makes it possible to use Twisted without any problems when fips is enabled. Also added GHA tests for the same.
```

Comment:
is this about fips being enabled or disabled?

Note that this is not a commit message.
This is a release notes.

This is to let Twisted users knnow about this change, and let them know if they should upgrade to latest version or not.

I still don't understand how MD5 works with FIPS enabled.

---------

Still the changes from this PR as very useful and can be used for future hardening or raising warnings when MD5 is used

Reply:
>is this about fips being enabled or disabled?

Twisted works the same way like it works now but when openssl fips is enabled in the host where Twisted is running.

How MD5 works when fips is enabled?

From: https://docs.python.org/3/library/hashlib.html
```
Changed in version 3.9: All hashlib constructors take a keyword-only argument usedforsecurity with default value True. A false value allows the use of insecure and blocked hashing algorithms in restricted environments. False indicates that the hashing algorithm is not used in a security context, e.g. as a non-cryptographic one-way compression function.
```

Openssl has a mechanism to momentarily disable fips and I'm using that option here.

>This is to let Twisted users knnow about this change, and let them know if they should upgrade to latest version or not.

Of course users should update to the latest version of Twisted if they are using Twisted in fips enabled environment.
Also, this (`usedforsecurity` flag) is available in `python3 >= 3.9`.

Can you please suggest proper release notes for this? I will add the same. Thanks.

Reply:
> Openssl has a mechanism to momentarily disable fips and I'm using that option here.

So when MD5 is used, at that exact moment, the FIPS mode is disabled.

So this make Twisted work with an OpenSSL library that has the FIPS mode disabled.

Is that correct?

I still don't understand how FIPS mode is enable :)

10 years ago I worked with OpenSSL FIPS certification and I remember that MD5 was not supposted to supported at all with FIPS enabled.
If your software was using MD5, you were not able to pass any FIPS certification.

Reply:
>So when MD5 is used, at that exact moment, the FIPS mode is disabled.

Yes but disabled only in the context of Twisted process not system wide.

> So this make Twisted work with an OpenSSL library that has the FIPS mode disabled.

Yes, this is a feature in python & openssl.

> I still don't understand how FIPS mode is enable :)

Check `openssl md5 <file-path` after installing `openssl-fips-provider` in a Photon 5.0 container and you will see that fips is enabled. In Photon OS installing this package enables openssl fips.
Check this spec: https://github.com/vmware/photon/blob/5.0/SPECS/openssl/openssl.spec#L189
Hope this clarifies your doubts. Sorry if I'm not understanding your question properly.

> If your software was using MD5, you were not able to pass any FIPS certification.

Fips certification is an entirely different game, it is a long process and takes a lot of time. And at present, certification doesn't happen for applications, you get certification for openssl in the system and make all the apps use the openssl libraries which is already certified. This way, everything will be fips compliant by default.

Reply:
@adiroiban & @glyph is there any thing more info or changes needed from my end? If not shall we merge this change? Thanks.

Reply:
@sshedi the way to do get it back into the queue is to use the magic comment "review please" as I have done below :)

Reply:
Noted, thanks :)